### PR TITLE
Stabilize more APIs for the 1.20.0 release

### PR DESCRIPTION
**Pull request by alexcrichton**
In addition to the few stabilizations that have already landed, this cleans up the remaining APIs that are in `final-comment-period` right now to be stable by the 1.20.0 release

### CODE REVIEW

**Comment by alexcrichton** [comment]
Note that I've explicitly omitted [`RangeArgument`](https://github.com/rust-lang/rust/issues/30877) as I believe there's still activity happening on that.

**Comment by alexcrichton** [comment]
(or @sfackler)

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -445,6 +444,8 @@ declare_features! (
     // Allows the definition of associated constants in `trait` or `impl`
     // blocks.
     (accepted, associated_consts, "1.20.0", Some(29646)),
+    // Usage of the `compile_error!` macro
+    (accepted, compile_error, "1.20.0", Some(40872)),
```
Comment (kennytm) [inline src/libsyntax/feature_gate.rs]:
```
[00:04:22] error[E0026]: struct `feature_gate::Features` does not have a field named `compile_error`
[00:04:22]     --> /checkout/src/libsyntax/ext/expand.rs:1044:35
[00:04:22]      |
[00:04:22] 1013 | |                      Some(&Features { $field: true, .. }) => true,
[00:04:22]      | |__________________________________________________^ struct `feature_gate::Features` does not have field `compile_error`
[00:04:22] ...
[00:04:22] 1034 | |/     feature_tests! {
[00:04:22] 1035 | ||         fn enable_quotes = quote,
[00:04:22] 1036 | ||         fn enable_asm = asm,
[00:04:22] 1037 | ||         fn enable_global_asm = global_asm,
[00:04:22] ...    ||
[00:04:22] 1044 |  |         fn enable_compile_error = compile_error,
[00:04:22]      |  |___________________________________^
[00:04:22] 1045 |  |     }
[00:04:22]      |  |_____- in this macro invocation
[00:04:22] 
[00:04:22] error: aborting due to previous error(s)
[00:04:22] 
[00:04:22] error: Could not compile `syntax`.
```

**Comment by aturon** [comment]
LGTM. Fire at will.

**Comment by kennytm** [comment]
libstd doc tests failed. 

```
[01:04:15] ---- ffi/c_str.rs - ffi::c_str::CStr::into_c_string (line 1003) stdout ----
[01:04:15] 	error: this feature has been stable since 1.20.0. Attribute no longer needed
[01:04:15]  --> <anon>:1:12
[01:04:15]   |
[01:04:15] 1 | #![feature(into_boxed_c_str)]
[01:04:15]   |            ^^^^^^^^^^^^^^^^
[01:04:15]   |
...
[01:04:15] failures:
[01:04:15]     ffi/c_str.rs - ffi::c_str::CStr::into_c_string (line 1003)
[01:04:15]     ffi/c_str.rs - ffi::c_str::CString::as_c_str (line 455)
[01:04:15]     ffi/c_str.rs - ffi::c_str::CString::into_boxed_c_str (line 476)
[01:04:15]     ffi/os_str.rs - ffi::os_str::OsString::into_boxed_os_str (line 254)
```

**Comment by kennytm** [comment]
`x86_64-gnu-aux`, #43453, spurious.

**Comment by Aaronepower** [comment]
@alexcrichton What does mean for `impl SliceIndex<str> for Range*` being stable when the `SliceIndex` trait is still unstable?

**Comment by alexcrichton** [comment]
@Aaronepower oh that's actually mostly irrelevant in terms of stabilization, the real stabilization  here is methods like `str::get` that consume these impls.

**Pull request merged**